### DISCUSSION TITLE
docker buildx offline?

### DISCUSSION THREAD

**Discussion by Will-NN** [docker/buildx#1088] [Q&A]
Hi all. First time using buildx here, I'm trying to build an arm64 image, and looking for some help.

I'm using docker 19.03 so I enabled the experimental mode to use buildx, and verified that I can do "docker buidx ...", etc.... Then I created a new builder "docker buildx create --name mybuilder" and set to use it "docker buildx use mybuilder". However, when I tried "docker build inspect --bootstrap", I got an error something like "ERROR [internal] booting buildkit". I realized that this error was due to the fact that I don't have an internet connection on my build machine. Once I connected my build machine to the internet, this command went through. Then, what is the process of building with buildx on an offline machine? (since I'll need to repeat the process on an offline machine later.)

**Comment by thaJeztah** [comment]
I guess creating the build will pull an image to run the builder in.

You should be able to run buildx against the version of buildkit embedded in the docker 19.03 daemon (but some features may not be available as it's an older version of buildkit)

Not sure if there's a way to create the container builders from an offline image

**Reply by Will-NN** [reply]
>> You should be able to run buildx against the version of buildkit embedded in the docker 19.03
How do I do this? I guess I'll have to specify the buildkit or somehow tell the builder to use the local buildkit?

**Reply by thaJeztah** [reply]
`docker buildx build` by default connects to `/var/run/docker.sock` (or the host that you're using for your docker commands); in that case it automatically uses the buildkit instance that's provided by the docker daemon

**Reply by Will-NN** [reply]
I guess I still don't understand it. No matter what I do, "docker buildx build" still trying to pull an image from the docker registry (which would fail once I disconnect my host from the internet)

[+] Building 0.1s (1/1) FINISHED                                                                             
 => ERROR [internal] booting buildkit                                                                   0.1s
 => => pulling image moby/buildkit:buildx-stable-1                                                      0.1s
------
 > [internal] booting buildkit:
------
error: Error response from daemon: Get https://registry-1.docker.io/v2/: dial tcp: lookup registry-1.docker.io

**Reply by morlay** [reply]
You could set image of your private container registry.

**Reply by ericwongwh** [reply]
@morlay, how to configure buildx to pull buildkit image from private container registry?

**Reply by bropuffshroom** [reply]
> @morlay, how to configure buildx to pull buildkit image from private container registry?

`
docker buildx create --name builder1 --driver=docker-container --driver-opt image=YOUR_PRIVATE_REGISTRY.io/buildkit:buildx-stable-1
`

you can refer to the documentation here https://docs.docker.com/build/drivers/docker-container/#usage

**Reply by nejtr0n** [reply]
I've got 403 error, when trying to pull image with buildkit driver-opt.

**Comment by disdorqin** [comment]
What you're seeing is expected. When you create a `docker-container` builder, Buildx pulls the BuildKit image (`moby/buildkit:buildx-stable-1`) from a registry the first time it boots — your error log shows that exact pull failing. So an offline host can't start that builder until the image is already present locally.

First, check what's already on the offline host:

```bash
docker buildx version
docker images
docker buildx ls
```

`docker images` tells you whether `moby/buildkit` is staged. If it isn't, stage it from a connected machine, then move it over:

```bash
docker pull moby/buildkit:buildx-stable-1
docker save moby/buildkit:buildx-stable-1 -o buildkit.tar
# copy buildkit.tar to the offline host, then:
docker load -i buildkit.tar
```

After that, `docker buildx create` won't need to reach the registry.

Two related options worth knowing:
- The default builder can use the BuildKit instance already embedded in the Docker daemon, so it doesn't need a separate image pull. For an offline setup you can just use the default builder instead of creating a container one.
- A container builder can pull from a local/private registry mirror if you stage the image there, via `--driver-opt image=<your-registry>/buildkit:buildx-stable-1`.

Note that `--network=none` only controls the build's own network — it does not replace BuildKit's one-time bootstrap pull.

If you can share `docker buildx version` and `docker images | grep -i buildkit` from the offline host, that shows whether the image is already staged.
